src/main/java/com/mixpanel/android/mpmetrics/MixpanelPushNotification.java

```diff
@@ -290,31 +289,193 @@ protected void maybeAddActionButtons() {
-                new Intent(Intent.ACTION_VIEW, Uri.parse(uri)),
-                PendingIntent.FLAG_UPDATE_CURRENT
-        );
+    protected PushTapAction buildOnTap(String onTapStr) {
+        PushTapAction onTap = null;
+        if (null != onTapStr) {
+            try {
+                final JSONObject onTapJSON = new JSONObject(onTapStr);
+                final String typeFromJSON = onTapJSON.getString("type");
+
+                if (!typeFromJSON.equals(PushTapTarget.HOMESCREEN.getTarget())) {
+                    final String uriFromJSON = onTapJSON.getString("uri");
+                    onTap = new PushTapAction(PushTapTarget.fromString(typeFromJSON), uriFromJSON);
+                } else {
+                    onTap = new PushTapAction(PushTapTarget.fromString(typeFromJSON));
+                }
+
+            } catch (JSONException e){
+                MPLog.d(LOGTAG, "Couldn't parse JSON Object for \'mp_ontap\'");
+                onTap = null;
+            }
+        }
+
+        return onTap;
+    }
+
+    protected PushTapAction buildOnTapFromURI(String uriString) {
+        PushTapAction onTap = null;
+
+        if (null != uriString) {
+            onTap = new PushTapAction(PushTapTarget.URL_IN_BROWSER, uriString);
+        }
+
+        return onTap;
+    }
+
+    protected PushTapAction getDefaultOnTap() {
+        return new PushTapAction(PushTapTarget.HOMESCREEN);
+    }
+
+    @TargetApi(20)
+    protected Notification.Action createAction(int icon, CharSequence title, PushTapAction onTap, String actionId, int index) {
+        return (new Notification.Action.Builder(icon, title, createActionIntent(onTap, actionId, title, index))).build();
+    }
+
+    protected PendingIntent createActionIntent(PushTapAction onTap, String buttonId, CharSequence label, int index) {
+        Intent routingIntent = getRoutingIntent(onTap, buttonId, label);
+        return PendingIntent.getActivity(context, ROUTING_REQUEST_CODE + index, routingIntent, PendingIntent.FLAG_CANCEL_CURRENT);
+    }
+
+    protected Intent getRoutingIntent(PushTapAction onTap, String buttonId, CharSequence label) {
+        Bundle options = buildBundle(onTap, buttonId, label);
+
+        Intent routingIntent = new Intent().
+                setClass(context, MixpanelNotificationRouteActivity.class).
+                putExtras(options).
+                setFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
+
+        verifyIntentPackage(routingIntent);
+        return routingIntent;
+    }
+
+    protected Intent getRoutingIntent(PushTapAction onTap) {
+        Bundle options = buildBundle(onTap);
+
+        Intent routingIntent = new Intent().
+                setClass(context, MixpanelNotificationRouteActivity.class).
+                putExtras(options).
+                setFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
+
+        verifyIntentPackage(routingIntent);
+        return routingIntent;
+    }
+
+    protected Bundle buildBundle(PushTapAction onTap) {
+        /**
```

Comment:
why doesn't this just call through to `buildBundle(PushTapAction onTap, String buttonId, CharSequence buttonLabel)`?